design/ceph-nfs-ganesha.md

```diff
@@ -0,0 +1,166 @@
+  initial export stored in the object that the server will start off exporting.
+  ```
+  # cat << EOF > ganesha-export.conf
+  EXPORT
+  {
+        Export_ID=100;
+        Protocols = 4;
+        Transports = TCP;
+        Path = /;
+        Pseudo = /cephfs/;
+        Access_Type = RW;
+        Attr_Expiration_Time = 0;
+        Delegations = R;
+        Squash = No_root_squash;
+        FSAL {
+                Name = CEPH;
+        }
+  }
+  EOF
+  # rados -p myfs-data0 put ganesha-export ganesha-export.conf
+  ```
+
+### Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha CRD will specify the following:
+
+- RADOS pool where the Ganesha exports and NFS client recovery data will be
+  stored.
+
+- RADOS object in the pool that will have the Ganesha exports
+
+- The number of active Ganesha servers
+
+- The placement of the Ganesha servers
+
+- The resource limits (memory, CPU) of the Ganesha pods
+
+Below is an example NFS-Ganesha CRD, `nfs-ganesha.yaml`
+
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: NFSGanesha
+metadata:
+  name: my-nfs
+  namespace: rook-ceph
+spec:
+  export:
+    # RADOS pool where Ganesha exports and NFS client recovery data is stored.
+    pool: myfs-data0
+    # RADOS object in the pool that has the Ganesha exports
+    object: ganesha-export
+  server:
+    # the number of active ganesha servers
+    active: 1
+    # where to run the nfs ganesha server
+    placement:
+    #  nodeAffinity:
+    #    requiredDuringSchedulingIgnoredDuringExecution:
+    #      nodeSelectorTerms:
+    #      - matchExpressions:
+    #        - key: role
+    #          operator: In
+    #          values:
+    #          - mds-node
+    #  tolerations:
+    #  - key: mds-node
+    #    operator: Exists
+    #  podAffinity:
+    #  podAntiAffinity:
+    # The requests and limits set here allow the ganesha pod(s) to use half of one CPU core and 1 gigabyte of memory
+    resources:
+    #  limits:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+    #  requests:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+```
+
+When the  nfs-ganesha.yaml is created the following will happen:
+
+- The Rook operator sees the creation of the NFS-Ganesha CRD and sets up the
+  specified Ganesha configuration.
+
+- A service will be created, which will provide the NFS clients with a stable
+  IP address for the Ganesha server.
+
+- A deployment will be started that will start the ganesha pod with the
+  specified configuration.
+
+The initial implementation will be able to orchestrate only a single active
+Ganesha server that is highly available. It will be enhanced to orchestrate
+multiple active Ganesha server clusters.
+
+### Dynamically adding and removing exports of running Ganesha servers
+
+This will involve fetching the existing ganesha exports from the RADOS
+object, updating them, and putting them back in the RADOS object, and then
+sending DBus signals to the ganesha pods to make the  ganesha servers refresh
+their exports.
```

Comment:
Note that dbus in containers is currently problematic (though we may be able to work around this by running our own dbus daemon or systemd inside the container). We could explore other sorts of IPC here too though. dbus is how this has been traditionally done in ganesha, but there's nothing magical about it.

Reply:
>  restart the ganesha pods if needed

If a new export requires restarting the pods, then that would not be desirable because the Ganeshas would have to do recovery with grace periods.

Reply:
Agreed, since there are side effects of restarting the pods it should be avoided. As a last resort, the operator would hopefully be smart enough (eventually) to handle when this does happen. For example, if Kubernetes decides to reschedule the Ganesha pod because a node went down, should the operator watch for this situation and assist with the recovery?

Reply:
There are two separate points here:
* where should the config info be stored? RADOS is one idea, but we could always generate config files directly within the container. I don't feel strongly about either way.

* How would we add/remove exports dynamically? Restarting ganesha is one way, but that can be disruptive for the whole cluster. There are dbus commands that can add and remove exports from a running server, so that's probably preferable.

As I understand it, dynamically added and removed exports also need to be persisted in the configuration so that they will be seen after reboot, so we'll need to do that either way.

FWIW, we may be able to run a dbus daemon within the container to handle this. That's an avenue that needs to be explored.

Reply:
Looking over the ganesha code, I think we avoid dbus here altogether. We can change the config file and then send it a SIGHUP to make it reread the config file. Not all options are changeable via this, but EXPORTS should be OK. Here's the comment header for reread_config():

```
/**
 * @brief Reread the configuration file to accomplish update of options.
 *
 * The following option blocks are currently supported for update:
 *
 * LOG {}
 * LOG { COMPONENTS {} }
 * LOG { FACILITY {} }
 * LOG { FORMAT {} }
 * EXPORT {}
 * EXPORT { CLIENT {} }
 *
 */
```

Reply:
@jtlayton , I remember now the reason for sticking to DBus to update exports with manila driver rather than SIGHUP.  With SIGHUP you can add exports but cannot remove exports. From,
https://github.com/nfs-ganesha/nfs-ganesha/wiki/Dbusinterface
"A specific export can be updated with dBus, and all exports in the config are refreshed on SIGHUP. With this, you can dynamically update everything except export_id, path, and pseudo, those basically imply removing the export and adding it again, so just use the dBus remove and add commands. SIGHUP will also add any new exports that have been added to the config file (it will not remove any exports - it doesn't make any check to see if any export was no longer present in the config file). The dBus update export will also add the export if the specified export does not already exist."

I guess we are stuck to using DBus?

Reply:
Maybe. Let me bring it up at this week's ganesha meeting. Maybe it would make sense to remove exports that are no longer defined.